src/integrations/prefect-dbt/prefect_dbt/core/runner.py

```diff
@@ -227,32 +227,45 @@ def _get_upstream_manifest_nodes_and_configs(
         self,
         manifest_node: ManifestNode,
     ) -> list[tuple[Union[ManifestNode, SourceDefinition], dict[str, Any]]]:
-        """Get upstream nodes for a given node"""
+        """
+        Get upstream nodes for a given node.
+        Ephemeral nodes are traversed recursively to find non-ephemeral dependencies.
+        """
         upstream_manifest_nodes: list[
             tuple[Union[ManifestNode, SourceDefinition], dict[str, Any]]
         ] = []
+        visited: set[str] = set()
 
-        for depends_on_node in manifest_node.depends_on_nodes:  # type: ignore[reportUnknownMemberType]
-            depends_manifest_node = self.manifest.nodes.get(
-                depends_on_node  # type: ignore[reportUnknownMemberType]
-            ) or self.manifest.sources.get(depends_on_node)  # type: ignore[reportUnknownMemberType]
+        def collect(node: ManifestNode):
```

Comment:
`depends_manifest_node` might be a `ManifestNode` or a `SourceDefinition` right?

```suggestion
        def collect(node: ManifestNode | SourceDefinition):
```

Reply:
Good catch. Thank you!